Preliminary representation of rvalue classical expressions in Terra (#30)
Qiskit/rfcs: 0030-simple-classical-representations.md

- On Terra, we need more of a feel for what will actually help us represent more complex classical constructs

## User Benefit

- Users of IBM dynamic circuits will be able to take advantage of the classical-processing capabilities of those compilers from within Qiskit
- This support is not limited to IBM so may extend further, but we are not aware of other dynamic-circuits hardware accessible through Qiskit
- For Terra development: this is a "build one to throw away"-type sytem, so we can see in actual usage what works and doesn't for us before we commit to further classical processing.

## Design Proposal

The intent of this RFC is to define an exploratory representation for a simple subset of classical bit-level operations, which expand the current state of Terra's `IfElseOp.condition`.
This representation is not required to be the base for future expansion; it will be permissible to introduce a completely different classical-value type system if that is more appropriate, and deprecate this one.
To that end, this proposal will not include adding any additional public methods to any existing Terra classes that might conflict with an alternative implementation.

> In particular, this proposal will not consider adding ease-of-use arithmetic methods to `Clbit` or `ClassicalRegister`, nor will it add expressions that can be used as the parameters to gates.

Author: I can add a couple of lines about that if you like, though there's equally no hard guarantee that it _will_ be removed; I was just designing such that it could be our standard methods of deprecation.

> In particular, this proposal will not consider adding ease-of-use arithmetic methods to `Clbit` or `ClassicalRegister`, nor will it add expressions that can be used as the parameters to gates.

Author: Added a note about not requiring this to be removed and a short note on deprecations in e6dd506.

### Representation

`IfElseOp.condition` and `WhileLoopOp.condition` will expand their allowed values to be an "expression" that evaluates to Boolean.
The allowed expressions will be the following:

- Object:
  - literal `bool` (`True` or `False`);
  - literal `uint[_]` (non-negative Python `int`);
  - runtime `bool` (`Clbit`);
  - runtime `uint[n]` (`ClassicalRegister(size=n)`).
- Unary:
  - `uint[n] -> uint[n]`: `~` (bitwise negation);

>   - `bool -> bool`: `!` (boolean negation).

Author: Oh, that was a dumb C/Rust/Python mixup.  I guess we'd have to rely on `~` for logical negation too.

>   - `bool -> bool`: `!` (boolean negation).

Author: Fixed in b8b7a19.

- Binary:
  - `uint[n] * uint[n] -> uint[n]`: `&`, `|`, `^`;
  - `bool * bool -> bool`: `&&`, `||`;
  - `uint[n] * uint[n] -> bool`: `==`, `!=`, `<`, `<=`, `>`, `>=`.

For typing in this initial draft, we will make an exact bijection:

- `ClassicalRegister(size=n)` <=> `uint[n]`

> - `Clbit` <=> `bool`

Author: `Clbit` won't be a `uint[1]` in this type system, nor I think should it be eligible for any implicit cast to `uint[1]`, but I will overload `&`, `|` and `^` to also act on `bool` too (I actually thought that's what I'd written, but apparently not).

When we add explicit casting, I'm fine having an explicit cast `bool -> uint[_]` defined by `true -> 1` and `false -> 0`, but having a special implicit cast from `bool -> uint[1]` I think is liable to cause more problems than it solves, and I think all the real benefit can be derived by just allowing the bitwise operators to act on them.

> - `Clbit` <=> `bool`

Reviewer: This makes sense to me, can you add a note on this to future extensions (and update to include the overloading for `Clbit`s)?

> - `Clbit` <=> `bool`

Author: Done in 041bf5d.

Python Booleans (`True` and `False`) will be used as the `bool` literal values, as they currently are in `Instruction.condition`.
Non-negative Python ints will be used as the `uint` literal values.
These do not have an inherent width.
The width of a literal `uint[_]` will be inferred and fixed to make a containing binary expression validly typed, and mixed-width operations will be forbidden.

> A binary expression in this representation _must_ contain at least one type of fixed width; it will be an error to attempt to use the runtime logic on two literals (this simplifies the implementation, and we won't do constant folding).

Reviewer: Is there something we can say here about how the width will be inferred? Is this behavior that's coming from OpenQASM3?

Reviewer: I'm not sure how to interpret "it will be an error to attempt to use the runtime logic on two literals" in this context. Can you explain a little more?

Author: When I wrote this, I was intending `Equal(Value(2), Value(2))` to be a type error on construction, because that expression would involve evaluating a binary op with both inputs having type `uint[UNKNOWN]`, and that's something I wanted to avoid having to define semantics for.

For the inference, this is what I meant:
```python
creg = Value(ClassicalRegister(3))  # type: uint[3]
two = Value(2)  # type: uint[UNKNOWN]
three = Value(3)  # type: uint[UNKNOWN]
five = Value(5)  # type: uint[UNKNOWN]

# This binary op causes the type of `two` to be inferred as `uint[3]` to match.
# This would recurse through any partially type objects.
add_two = Add(creg, two)

# This expression is not valid, because it's a binop on two partial types.
eight = Add(three, five)
```

My intent was just to forbid this kind of thing for now in the actual expression tree, rather than needing to define semantics for how we handle intermediate numeric values of unknown width.  (For example, at what precision would we choose to handle an expression `312890318904839189082032 < 43819432893121004381904`?)

Author: Yeah, that would raise or warn becuase `0b1010` can't fit in a `uint[3]`.

This is not intended to be a complete set of classical expressions, for simplicity in this initial implementation.
Notably, this MVP will _not_ include:

1. arithmetic on `uint[n]` (`+`, `*`, etc);
2. indexing the bits resulting from one of these rvalue expresions;
3. slicing a `uint[n]` to retrieve a subset of its bits;
4. any method of assigning one of these expressions for reuse elsewhere in the circuit.

Points 1 through 3 are rejected in this initial proposal for the sake of keeping the number of expression-tree nodes and type-system entries small.
Point 4 is rejected because it would require the definition of additional `Operation` objects and potentially modifications to the `QuantumCircuit` class that we are not yet prepared to make.
All of these points are eligible to be included in future work, just not the MVP.

### Construction

#### Stage 1

> This is a temporary representation, so it is a non-goal to produce the best possible UX for constructing these objects.

Reviewer: ```suggestion
This is an experimental representation, so it is a non-goal to produce the best possible UX for constructing these objects.
```
Maybe?

In the initial representation phase, users will be required to build the expression-tree representations themselves.
This will look _something_ like (names up for discussion):

```python
from qiskit.circuit import Clbit, ClassicalRegister
from qiskit.circuit.classical import expr

loose = [Clbit() for _ in [None]*3]
cr1 = ClassicalRegister(2)
cr2 = ClassicalRegister(2)

# Equivalent to the current condition `(cr, 2)`:
equal_reg = expr.Equal(expr.Value(cr1), expr.Value(2))

# Equivalent to the current condition `(loose[0], False)`
equal_bit = expr.Equal(expr.Value(loose[0]), expr.Value(False))

# The above example, but using a Bool-valued unary expression:
not_bit = expr.Not(expr.Value(loose[0]))

# The comparison 0 < cr1 <= cr2
bounded_reg = expr.And(
    expr.Less(expr.Value(0), expr.Value(cr1),
    expr.LessEqual(expr.Value(cr1), expr.Value(cr2),
)
```

> These expressions will then be given directly to `IfElseOp` or `WhileLoopOp` in their `condition` fields.

Reviewer: For sake of clarity, it might help to show these examples in the context of building an `IfElseOp` or `WhileLoopOp` since it seems a more common usage than building expressions separately.

Author: Done in 7f5424e.

These expressions will not be valid in the general `Instruction.condition` field; we do not want to expand any support for something we are already moving to remove.

#### Stage 2

Looking beyond the initial phase, it will be convenient to define a less verbose way for users to create these expressions.
Since this document is intended to define a representation that can be removed and deprecated from Terra easily if it needs to evolve, we will _not_ consider overloading the Python magic-methods `__and__` etc on `Clbit` and `ClassicalRegister` in general Terra usage.
However, we can do so in a limited scope, which will temporarily change the semantics.

> This will look like (continued from above code block, and the same examples in the same order):

Reviewer: I think it's worth evaluating how much this makes the deprecation and eventual removal path smoother for users vs. the near time usability cost. It seems like most usage of `IfElseOp` and `WhileLoopOp` will want a `with expr.build():` block, and (with the exception of `__eq__`) these magic methods aren't currently being used, and so could be deprecated as needed when used.

Author: If we choose to keep these objects, we can also just move the magic methods onto the core objects, but that doesn't really feel great to me; I think magic methods should always feel _obvious_ about what they're going to do, and detached from a builder context, I don't think "`Clbit == Clbit` makes a runtime expression" is super obvious.

```python
with expr.build():
    equal_reg = cr1 == 2
    equal_bit = loose[0] == False
    not_bit = ~loose[0]
    bounded_reg = (0 < cr1) & (cr1 <= cr2)
```

> It is not possible to overload the behaviour of the Python Boolean operators (`not`, `and` and `or`) for classes, which is why the bitwise operations `~` and `&` are used instead (with the corresponding precedence frustrations).

Reviewer: I could see this being a source of confusion. e.g. "Why can I `cr[0] == cr[1]` but not `cr[0] and cr[1]`?". Are there any cases where this could lead to errors like "I wanted a bitwise AND but got a logical AND"?

Author: The way I wrote this, you'll always have the option to use the regular constructors, or potentially Kevin's suggested chained methods, even in the middle of the builder context.  But yeah, it's not ideal, but imo magic methods give the best expression legibility in the majority of cases, so until we decide to commit to a _complete_ function-based builder interface where we can do AST inspection, this is my preferred choice.

Also: my intent was for `&` to still be bitwise AND, it's just also defined on bools.  You'd have to be explicit to get logical operations (with the corresponding short-circuiting)

Within the `expr.build()` context, the magic methods of `ClassicalRegister` and `Clbit` will be temporarily overridden to be an eager builder interface for the syntax tree given above.

> This means that the expressions will not need to be bound to variables in order to use them, as the resulting objects will be well-defined and the correct tree form.

Reviewer: I agree with the statement here, but I'm not sure how it follows from the sentence above.

Author: I was trying to say that you don't need to do
```python
with expr.build():
   my_cond = creg1 == creg2
with qc.if_test(my_cond):
   pass
```
(i.e. you don't need the `build` context to end), you'd be able to directly build the circuit as
```python
with expr.build():
    with qc.if_test(creg1 == creg2):
        pass
```
and the `if_test` method would see the correct objects at the point of call.

For example, it will be valid to do

```python
qc = QuantumCircuit([Qubit()], loose)
with expr.build():
    with qc.if_test(~loose[0]):
        qc.h(0)
```

The reason to require entering the builder context to make the magic methods available is for potential future ease of change/removal/deprecation.
The mutation of global state required to do this makes the builder interface non-thread-safe, but this is not judged to be a problem; it is uncommon to attempt to multithread during manual circuit construction, and the advantages of easy removal/representation change completely outweigh the potential concern.

## Detailed Design

> The new functionality will live in in `qiskit.circuit.classical`.

Reviewer 2: ```suggestion
The new functionality will live in `qiskit.circuit.classical`.
```

The expression-builder objects will all be exposed from `qiskit.circuit.classical.expr`, to allow people to bring the desired values into scope either with a namespace prefix `expr` or by a star-import of limited scope.

> From this point on, I will omit the prefix `qiskit.circuit.classical.`.

The class `expr.Expression` will be an abstract base class with two read-only fields:
```python
# file: qiskit/circuit/classical/expr.py

class Expression(ABC):
    resources: typing.Set[Clbit]
    type: qiskit.circuit.classical.Type
```

> The `resources` field is effectively a `use-def` tracker, which will more easily enable the conversion to data-flow (`DAGCircuit`) format to evaluate the necessary wires of an instruction.

Reviewer: Not a compiler expert by any means, but I think this is not exactly the same as `use-def` tracking, which makes some assurances about the value of a variable not being re-defined between definition and use. The `Clbit`s in `QuantumCircuit` now follow more like memory semantics, so there are no guarantees of their value or when they were last modified.

Author: Yeah, you're right.  In retrospect, I'm not even going to have the `resources` field anyway - I'll just make an `ExpressionVisitor` subclass that calculates them.  There's no reason that these "resources" should be any more special than any other shared property.

Author: Fixed in 2f80715.

The `type` field is a resolved type of the contained expression.

> In this initial release, the only time that `type` will not be a fully qualified type will be when representing a `expr.Value` of a `uint` literal.

Reviewer 2: Not fully-qualified in that the size isn't fixed?

Author: yeah, "qualified" doesn't seem like the right word here.  I'm pretty sure there is a correct word for it, I just can't think of it.

Type errors will raise immediately on attempted construction.
In the future, we may consider relaxing this restriction and having a type-evaluation pass that runs on the complete `QuantumCircuit` AST representation, but that is for later work.

The expressions in the above examples will be final (uninheritable) subclasses of `Expression`:
```python
class Value(Expression):
    value: Any

class Less(Expression):
    left: Expression
    right: Expression

...
```

Similarly, the type system will be represented by values of the type `types.Type`:
```python
# file: qiskit/circuit/classical/types.py

class Type(ABC):
    pass

@typing.final
class Bool(Type):
    pass

UNKNOWN = object()

@typing.final
class Uint(Type):
    size: int | Literal[UNKNOWN]
```

## Implementation plan

PR 1:
- write representation of `Expression` and `Type`
- test that these objects can be constructed and handle type checking correctly

PR 2 (depends on 1):
- add support to `Expression` of evaluated type `Bool` to `IfElseOp.condition` and `WhileLoopOp.condition`

PR 3 (depends on 2):

> - add support for exporting these `Expression` conditions to the OpenQASM 3 

Reviewer 2: ```suggestion
- add support for exporting these `Expression` conditions to OpenQASM 3 
```

# Preliminary representation of rvalue classical expression in Qiskit

> | **Status**        | **Proposed/Accepted/Deprecated** |
> |:------------------|:---------------------------------------------|
> | **RFC #**         | ####                                         |

Reviewer 3: ```suggestion
| **Status**        | **Accepted** |
|:------------------|:---------------------------------------------|
| **RFC #**         |  30                                        |
```

In particular, this proposal will not consider adding ease-of-use arithmetic methods to `Clbit` or `ClassicalRegister`, nor will it add expressions that can be used as the parameters to gates.

Notes:

- We want to keep a level of interoperability with the OpenQASM 3 description of hybrid quantum–classical programs, to have better cross-ecosystem compatibility.
  When in doubt, we probably want to run towards the design choices made there.
- There is no _requirement_ to throw this representation away again.
  If it works, we can build on it and keep it.

### Representation

`IfElseOp.condition` and `WhileLoopOp.condition` will expand their allowed values to be an "expression" that evaluates to Boolean.
The allowed expressions will be the following:

- Object:
  - literal `bool` (`True` or `False`);
  - literal `uint[_]` (non-negative Python `int`);
  - runtime `bool` (`Clbit`);
  - runtime `uint[n]` (`ClassicalRegister(size=n)`).
- Unary:
  - `uint[n] -> uint[n]`: bitwise negation
  - `bool -> bool`: Boolean negation
- Binary:

>   - `uint[n] * uint[n] -> uint[n]`: `&`, `|`, `^`;

Reviewer 4: Should bit shift `<<` and `>>` be allowed as well?

Author: For sure, in the future - this is just meant to be an MVP and I didn't want the available expression tree to be huge for a first attempt.

Reviewer: I think given that these are already defined in OpenQASM3 (and available on IBM hardware), and open the door to some applications, it would be good to have these in the first MVP (unless there's a subtlety that would make them more complex to implement than those already listed).

Reviewer: That's a good point, and agree that's a good approach for now.
